src/parsing/lpParser.mly

```diff
@@ -242,7 +242,17 @@ proof_end:
   | ADMITTED { make_pos $sloc Syntax.P_proof_admitted }
   | END { make_pos $sloc Syntax.P_proof_end }
 
-proof: BEGIN ts=terminated(tactic, SEMICOLON)* pe=proof_end { ts, pe }
+proof:
+  (*in case there are more than one goal to prove in the beginning*)
+  | BEGIN l=subproof+ pe=proof_end { l, pe }
+  (*in case there is just one goal to prove in the beginning*)
+  | BEGIN l=subproofbis pe=proof_end { [l], pe }
+
+subproof: L_CU_BRACKET l=terminated(proof_step, SEMICOLON)* R_CU_BRACKET { l }
+
+subproofbis: l=terminated(proof_step, SEMICOLON)* { l }
```

Comment:
We can get rid of the last semicolon and do some factorization by having something like:
```
proof:
  | BEGIN l=subproof+ pe=proof_end { l, pe }
  | BEGIN l=proof_steps pe=proof_end { [l], pe }
subproof: L_CU_BRACKET l=proof_steps R_CU_BRACKET { l }
proof_steps: l=separated_non_empty_list(SEMICOLON, proof_step) { l }
```

Reply:
My suggestion :
```menhir
proof:
  (*in case there are more than one goal to prove in the beginning*)
  | BEGIN l=braced_proof_steps+ pe=proof_end { l, pe }
  (*in case there is just one goal to prove in the beginning*)
  | BEGIN l=separated_non_empty_list(SEMICOLON, proof_step) SEMICOLON? pe=proof_end { [l], pe }

braced_proof_steps: L_CU_BRACKET l=separated_non_empty_list(SEMICOLON, proof_step) SEMICOLON? R_CU_BRACKET { l }
proof_step: t=tactic l=braced_proof_steps* { Tactic(t, l) }
```

Last ";" in blocks would be optional (but I think they should be allowed, like in OCaml).
Maybe a dummy "split" tactic to introduce subproofs/braced_proof_steps at the toplevel would make the AST more simple ?
Also, I usually rename constructions like "separated_non_empty_list" to make them more usable, and use Menhir's string aliases for tokens so maybe my suggestion isn't the best here

Reply:
My suggestion doesn't work (shift/reduce conflict), so here goes one that works

```menhir
nonempty_sep_except_maybe_last(arg, sep):
  | a=arg { [a] }
  | a=arg sep { [a] }
  | a=arg sep l=nonempty_sep_except_maybe_last(arg, sep) { a :: l }


[...]

proof:
  (*in case there are more than one goal to prove in the beginning*)
  | BEGIN l=proof_block+ pe=proof_end { l, pe }
  (*in case there is just one goal to prove in the beginning*)
  | BEGIN l=nonempty_sep_except_maybe_last(proof_step, SEMICOLON) pe=proof_end { [l], pe }

proof_block: L_CU_BRACKET l=nonempty_sep_except_maybe_last(proof_step, SEMICOLON) R_CU_BRACKET { l }
proof_step: t=tactic l=proof_block* { Tactic(t, l) }
```

A conflict remains, because curly braces are also used for explicit arguments, and using TYPE_QUERY they can be mixed up